dataproxy/service.go

```diff
@@ -0,0 +1,116 @@
+package dataproxy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/ioutils"
+
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/flyteorg/flyteadmin/pkg/config"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/flyteorg/flytestdlib/storage"
+	"github.com/flyteorg/stow"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+)
+
+type Service struct {
+	service.DataProxyServer
+	*service.UnimplementedDataProxyServer
+
+	cfg           config.DataProxyConfig
+	dataStore     *storage.DataStore
+	shardSelector ioutils.ShardSelector
+}
+
+// CreateUploadLocation creates a temporary signed url to allow callers to upload content.
+func (s Service) CreateUploadLocation(ctx context.Context, req *service.CreateUploadLocationRequest) (
+	*service.CreateUploadLocationResponse, error) {
+
+	if len(req.Project) == 0 || len(req.Domain) == 0 {
+		return nil, fmt.Errorf("prjoect and domain are required parameters")
+	}
+
+	if expiresIn := req.ExpiresIn; expiresIn != nil {
+		if !expiresIn.IsValid() {
+			return nil, fmt.Errorf("expiresIn [%v] is invalid", expiresIn)
+		}
+
+		if expiresIn.AsDuration() > s.cfg.Upload.MaxExpiresIn.Duration {
+			return nil, fmt.Errorf("expiresIn [%v] cannot exceed max allowed expiration [%v]",
+				expiresIn.AsDuration().String(), s.cfg.Upload.MaxExpiresIn.String())
+		}
+	} else {
+		req.ExpiresIn = durationpb.New(s.cfg.Upload.MaxExpiresIn.Duration)
+	}
+
+	if len(req.Suffix) == 0 {
+		req.Suffix = rand.String(20)
+	}
+
+	storagePath, err := createShardedStorageLocation(ctx, req, s.shardSelector, s.dataStore, s.cfg.Upload)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.dataStore.CreateSignedURL(ctx, storagePath, storage.SignedURLProperties{
+		Scope:     stow.ClientMethodPut,
+		ExpiresIn: req.ExpiresIn.AsDuration(),
+		// TODO: pass max allowed upload size
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a signed url. Error: %w", err)
+	}
+
+	return &service.CreateUploadLocationResponse{
+		SignedUrl: resp.URL.String(),
+		NativeUrl: storagePath.String(),
+		ExpiresAt: timestamppb.New(time.Now().Add(req.ExpiresIn.AsDuration())),
+	}, nil
+}
+
+func createShardedStorageLocation(ctx context.Context, req *service.CreateUploadLocationRequest,
```

Comment:
interesting, does this logic not currently exist elsewhere?

Reply:
This is the closest to this logic: https://github.com/flyteorg/flyteplugins/blob/60b94c688ef2b98aa53a9224b529ac672af04540/go/tasks/pluginmachinery/ioutils/raw_output_path.go#L55-L68